gen/constant.go

```diff
@@ -20,17 +20,222 @@
+		if !ok {
+			return "", fmt.Errorf(
+				"cannot cast struct %v to %v", v, t.ThriftName())
+		}
+		return g.TextTemplate(
+			`
+			<$fields := .Spec.Fields>
+			&<typeName .Spec>{
+				<range $name, $value := .Value.Fields>
+					<$field := $fields.FindByName $name>
+					<if and (not $field.Required) (isPrimitiveType $field.Type)>
+						<goCase $field.Name>: <primitiveValueRef $value $field.Type>,
+					<else>
+						<goCase $field.Name>: <constantValue $value $field.Type>,
+					<end>
+				<end>
+			}`, struct {
+				Spec  *compile.StructSpec
+				Value *compile.ConstantStruct
+			}{Spec: s, Value: v},
+			TemplateFunc("constantValue", ConstantValue),
+			TemplateFunc("primitiveValueRef", primitiveValueRef),
+		)
+	case compile.EnumItemReference:
+		return g.TextTemplate(`<typeName .Enum><goCase .Item.Name>`, v)
+	case compile.ConstReference:
+		return g.LookupConstantName(v.Target)
+	default:
+		panic(fmt.Sprintf("Unknown constant value %v (%T)", c, c))
+	}
+}
+
+// helper to generate pointers to primitives
+func primitiveValueRef(g Generator, c compile.ConstantValue, t compile.TypeSpec) (string, error) {
+	var toptr string
+	switch t {
+	case compile.BoolSpec:
+		toptr = "_boolptr"
+		err := g.EnsureDeclared(
+			`func _boolptr(v bool) *bool { return &v }`, nil)
+		if err != nil {
+			return "", err
+		}
+	case compile.I8Spec:
+		toptr = "_i8ptr"
+		err := g.EnsureDeclared(
+			`func _i8ptr(v int8) *int8 { return &v }`, nil)
+		if err != nil {
+			return "", err
+		}
+	case compile.I16Spec:
+		toptr = "_i16ptr"
+		err := g.EnsureDeclared(
+			`func _i16ptr(v int16) *int16 { return &v }`, nil)
+		if err != nil {
+			return "", err
+		}
+	case compile.I32Spec:
+		toptr = "_i32ptr"
+		err := g.EnsureDeclared(
+			`func _i32ptr(v int32) *int32 { return &v }`, nil)
+		if err != nil {
+			return "", err
+		}
+	case compile.I64Spec:
+		toptr = "_i64ptr"
+		err := g.EnsureDeclared(
+			`func _i64ptr(v int64) *int64 { return &v }`, nil)
+		if err != nil {
+			return "", err
+		}
+	case compile.DoubleSpec:
+		toptr = "_doubleptr"
+		err := g.EnsureDeclared(
+			`func _doubleptr(v float64) *float64 { return &v }`, nil)
+		if err != nil {
+			return "", err
+		}
+	case compile.StringSpec:
+		toptr = "_stringptr"
+		err := g.EnsureDeclared(
+			`func _stringptr(v string) *string { return &v }`, nil)
+		if err != nil {
+			return "", err
+		}
+	default:
+		// fall through
+	}
+
+	switch t.(type) {
+	case *compile.EnumSpec, *compile.TypedefSpec:
+		toptr = fmt.Sprintf("_%v_ptr", t.ThriftName())
+		err := g.EnsureDeclared(
+			`func _<.ThriftName>_ptr(v <typeReference .>) *<typeReference .> {
+				return &v
+			}`, t)
+		if err != nil {
+			return "", err
+		}
+	default:
+		// fall through
+	}
+
+	if toptr == "" {
+		panic(fmt.Sprintf(
+			"primitiveValueRef called with %v which is not a primitive",
+			t.ThriftName()))
+	}
+
+	s, err := ConstantValue(g, c, t)
+	if err == nil {
+		s = fmt.Sprintf("%v(%v)", toptr, s)
+	}
+	return s, err
```

Comment:
do we want to return the string even if there's an error?
why not write this as
```go
s, err := ConstantValue(..)
if err != nil {
  return "", err
}

return fmt.Sprintf(...)
```